### PR TITLE
fix(CI): surface cluster create/destroy events

### PR DESCRIPTION
**Pull request by gavin-stackrox**
## Description

This PR improves how cluster create & destroy failures are surfaced in CI. Using related changes in automation-flavors, these events result in junit records which then feed into the junit -> JIRA -> BigQuery & Slack pipeline.

Changes include:

- Generate "Cluster" junit based on the state shared from automation-flavors and GKE.
- Allow PRs to run the junit pipeline if they are labeled "ci-test-junit-processing".
  - JIRAs are created in the RS project.
  - Slack goes to #acs-slack-ci-integration-testing.
 - Refactored the Slack send function.
 - Added skipped junit support.

## Checklist
- [x] Investigated and inspected CI test results
- [x] Update junit2jira with a released version

## Testing

While the changes are shared between test suites and cluster flavors, for
confidence I plan on covering:

- GKE & OCP
- UI, Non-Groovy, QA and Operator

As part of testing, ensure that details presented in Junit, Slack, JIRA and
BigQuery are correct and useful. Links are correct, etc.

- JIRA: https://issues.redhat.com/issues/?jql=project%20%3D%20RS
- Slack: https://redhat.enterprise.slack.com/archives/C068WMY4SEQ

### Happy path

Clusters create & destroy OK, no tests fail. See run for `aa47857`.

- [x] - Cluster create & destroy success junits created
- [x] - no JIRA
- [x] - no Slack
- [x] - BQ cluster create and destroy pass records

### Tests fail

Clusters create & destroy OK, some tests fail. See run for `159455e`.

- [X] - Cluster create & destroy success junits created
- [X] - JIRA for failed test(s)
- [X] - Slack for failed tests(s)
- [X] - BQ cluster create and destroy have pass records, tests have some fail records

### Cluster creates fail

Clusters create failure, test are not run. See runs for `c734ba6`.

- [x] - Cluster create & (possibly) destroy failed junits created
- [x] - JIRA for failed cluster operation
- [x] - Slack for failed cluster operation
- [x] - BQ cluster create and (possibly) destroy fail records

Junit, JIRA and Slack should contain useful output to indicate the failure.

### Cluster destroys fail

Clusters destroy failure - difficult to force without a total test meltdown i.e.
manually destroy during test so that the process destroy fails.

#### Handling of jobs that fail without junit failure records

e.g. all tests are success but a post step blows up without a junit. Expect an error message to Slack indicating that CI failure handling could use improvement.

- [x] https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/stackrox_stackrox/8958/pull-ci-stackrox-stackrox-master-gke-nongroovy-e2e-tests/1734648970644295680 a failure in post test resulted in no junit and this slack: https://redhat-internal.slack.com/archives/C068WMY4SEQ/p1702411712519609

#### Un-parseable junit handling

- [x] junit2jira does not create slack attachments as expected. Previously these would surface in Slack as https://redhat-internal.slack.com/archives/CLUNQEEMA/p1701684985680249. See `519e478` run. An error message is sent to slack as expected https://redhat-internal.slack.com/archives/C068WMY4SEQ/p1702359791403429.

## Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

File: scripts/ci/lib.sh [hunk 1]
```diff
@@ -898,20 +896,27 @@ get_pr_details() {
...
     headers=()
     if [[ -n "${GITHUB_TOKEN:-}" ]]; then
         headers+=(-H "Authorization: token ${GITHUB_TOKEN}")
     fi
 
     url="https://api.github.com/repos/${org}/${repo}/pulls/${pull_request}"
-    pr_details=$(curl --retry 5 -sS "${headers[@]}" "${url}")
+
+    if ! pr_details=$(curl --retry 5 -sS "${headers[@]}" "${url}"); then
+        echo "Github API error: $pr_details/$?" >&2
```
Comment (porridge) [inline scripts/ci/lib.sh:913]:
```suggestion
        echo "Github API error: $pr_details, exit code $?" >&2
```

File: .openshift-ci/end.sh [hunk 2]
```diff
@@ -8,63 +8,154 @@ source "$ROOT/scripts/ci/lib.sh"
...
-        ;;
-    *)
-        info "ERROR: unexpected state in end.sh: ${combined}"
-        outcome="${OUTCOME_FAILED}"
-        ;;
-esac
-
-update_job_record outcome "${outcome}" stopped_at "CURRENT_TIMESTAMP()"
+end() {
+    info "And it shall end"
```
Comment (porridge) [inline .openshift-ci/end.sh:12]: While at it, can you please improve this message? I'd have no idea what it means. What is "it" in this context?

File: .openshift-ci/end.sh [hunk 3]
```diff
@@ -8,63 +8,154 @@ source "$ROOT/scripts/ci/lib.sh"
...
+
+    post_process_test_results "${END_SLACK_FAILURE_ATTACHMENTS}"
+
+    if [[ "${OVERALL_JOB_OUTCOME}" == "${OUTCOME_FAILED}" ]]; then
+        (send_slack_failure_summary) || { echo "ERROR: Could not slack a test failure message"; }
+    fi
+}
+
+determine_an_overall_job_outcome() {
+    # Determin a useful overall job outcome based on state shared from prior steps.
```
Comment (porridge) [inline .openshift-ci/end.sh:36]:
```suggestion
    # Determine a useful overall job outcome based on state shared from prior steps.
```

File: .openshift-ci/end.sh [hunk 4]
```diff
@@ -8,63 +8,154 @@ source "$ROOT/scripts/ci/lib.sh"
...
+    determine_an_overall_job_outcome
+
+    generate_cluster_junit
+
+    update_job_record outcome "${OVERALL_JOB_OUTCOME}" stopped_at "CURRENT_TIMESTAMP()"
+
+    post_process_test_results "${END_SLACK_FAILURE_ATTACHMENTS}"
+
+    if [[ "${OVERALL_JOB_OUTCOME}" == "${OUTCOME_FAILED}" ]]; then
+        (send_slack_failure_summary) || { echo "ERROR: Could not slack a test failure message"; }
```
Comment (porridge) [inline .openshift-ci/end.sh:31]: Should this echo go to stderr instead?

File: .openshift-ci/end.sh [hunk 5]
```diff
@@ -8,63 +8,154 @@ source "$ROOT/scripts/ci/lib.sh"
...
+_JUNIT_CLUSTER_CREATE_DESCRIPTION="Create"
+_JUNIT_CLUSTER_DESTROY_DESCRIPTION="Destroy"
+
+# generate_cluster_junit() - generate junit records for cluster create & destroy
+# pass & fail.
+generate_cluster_junit() {
+    # Change the presentation for cluster types
+    local cluster_flavor_variant="${CLUSTER_FLAVOR_VARIANT:-unknown}"
+    if [[ ${#cluster_flavor_variant} -le 5 ]]; then
+        # uppercase GKE, ARO, ROSA, etc
```
Comment (porridge) [inline .openshift-ci/end.sh:99]: Let's avoid clever shortcuts like this and make this a case/esac that explicitly looks at the prefix.

File: .openshift-ci/end.sh [hunk 6]
```diff
@@ -8,63 +8,154 @@ source "$ROOT/scripts/ci/lib.sh"
...
+        cluster_flavor_variant="${cluster_flavor_variant^^}"
```
Comment (porridge) [inline .openshift-ci/end.sh:100]: Please add a word of comment for those like myself who don't know the meaning of `^^` by heart :sweat_smile:

File: .openshift-ci/end.sh [hunk 7]
```diff
@@ -8,63 +8,154 @@ source "$ROOT/scripts/ci/lib.sh"
...
+    if [[ "${DESTROY_CLUSTER_OUTCOME:-}" == "${OUTCOME_PASSED}" ]]; then
+        save_junit_success "${_JUNIT_CLUSTER_CLASS}" "${_JUNIT_CLUSTER_DESTROY_DESCRIPTION} ${cluster_flavor_variant}"
+    elif [[ "${DESTROY_CLUSTER_OUTCOME:-}" == "${OUTCOME_FAILED}" && "${CREATE_CLUSTER_OUTCOME:-}" == "${OUTCOME_PASSED}" ]]; then
+        # Only record cluster destroy failures when the create succeeded. Otherwise it is mostly noise.
+        save_junit_failure "${_JUNIT_CLUSTER_CLASS}" "${_JUNIT_CLUSTER_DESTROY_DESCRIPTION} ${cluster_flavor_variant}" \
+            "${cluster_destroy_debug}"
+    fi
+}
+
+_NUM_LINE_OF_INTEREST=10
```
Comment (porridge) [inline .openshift-ci/end.sh:135]:
```suggestion
_NUM_LINES_OF_INTEREST=10
```

File: .openshift-ci/end.sh [hunk 8]
```diff
@@ -8,63 +8,154 @@ source "$ROOT/scripts/ci/lib.sh"
...
+
+summarize_cluster_debug() {
+    file="$1"
+
+    last_lines="$(tail --lines="${_NUM_LINE_OF_INTEREST}" "${file}")"
+    if [[ -z "${last_lines}" ]]; then
+        last_lines="No debug output found."
+    fi
+
+    suspicious_lines="$(grep -E 'error|warn|fatal|fail' "${file}" | tail --lines="${_NUM_LINE_OF_INTEREST}")"
```
Comment (porridge) [inline .openshift-ci/end.sh:145]:
```suggestion
    regex='error|warn|fatal|fail'
    suspicious_lines="$(grep -E "${regex}" "${file}" | tail --lines="${_NUM_LINE_OF_INTEREST}")"
```
And reuse the regex in the message below.

File: .openshift-ci/end.sh [hunk 9]
```diff
@@ -8,63 +8,154 @@ source "$ROOT/scripts/ci/lib.sh"
...
+    fi
+}
+
+_NUM_LINE_OF_INTEREST=10
+
+summarize_cluster_debug() {
+    file="$1"
+
+    last_lines="$(tail --lines="${_NUM_LINE_OF_INTEREST}" "${file}")"
+    if [[ -z "${last_lines}" ]]; then
```
Comment (porridge) [inline .openshift-ci/end.sh:141]:
```suggestion
    if [[ -z ${last_lines} ]]; then
```
`[[` does not need quotes and some linters complain about them

File: scripts/ci/lib.sh [hunk 10]
```diff
@@ -794,7 +791,7 @@ pr_has_label() {
     local exitstatus=0
     pr_details="${2:-$(get_pr_details)}" || exitstatus="$?"
     if [[ "$exitstatus" != "0" ]]; then
-        info "Warning: checking for a label in a non PR context"
+        info "Warning: checking for a label in a non PR context [${pr_details}/${exitstatus}]"
```
Comment (porridge) [inline scripts/ci/lib.sh:794]:
```suggestion
        info "Warning: checking for a label in a non PR context failed: ${pr_details}, exit code ${exitstatus}"
```

File: scripts/ci/lib.sh [hunk 11]
```diff
@@ -1140,40 +1149,63 @@ post_process_test_results() {
...
-            info "Converting JUNIT found in ${ARTIFACT_DIR} to CSV"
+        info "Post processing junit records to JIRA issues, BigQuery metrics and Slack attachments as appropriate"
+
+        prow_job_link="$(make_prow_job_link)"
+
+        if is_in_PR_context; then
+            if pr_has_label "ci-test-junit-processing"; then
+                create_jiras="true"
+            else
+                create_jiras="false"
```
Comment (porridge) [inline scripts/ci/lib.sh:1171]: Thinking about this, perhaps we _should_ report failures related to cluster create/teardown problems, even for PRs? For those rare PRs which tinker with cluster creation/teardown logic (and are somewhat expected to cause a failure in that area), we could have a label that disables reporting? :thinking:

**Review by porridge — APPROVED** [review]
LGTM overall, however some general thoughts below and inline, along with technical nitpicks.

This is a general improvement, thanks for breaking huge functions into chunks!

In general I think we're approaching the limits of what can be achieved in bash without going insane, maybe we should think about rewriting parts of this in a less controversial language?

**Comment by gavin-stackrox** [comment]
> LGTM overall, however some general thoughts below and inline, along with technical nitpicks.

Thanks as always @porridge for a thorough review.

> This is a general improvement, thanks for breaking huge functions into chunks!
> 
> In general I think we're approaching the limits of what can be achieved in bash without going insane, maybe we should think about rewriting parts of this in a less controversial language?

Leaving aside questions of sanity, I agree. There are quite a few helper functions/tools such as this that could be migrated to golang.

File: scripts/ci/lib.sh [hunk 12]
```diff
@@ -1410,30 +1516,32 @@ get_junit_misc_dir() {
     echo "${ARTIFACT_DIR}/junit-misc"
 }
 
+_JUNIT_DISPOSITION_SUCCESS="SUCCESS"
```
[the remark this answers is not in the document]
Comment (gavin-stackrox) [reply scripts/ci/lib.sh:1519]: I'm slightly reluctant to deprive my colleagues of a learning opportunity but I'll do it.

File: scripts/ci/lib.sh [hunk 13]
```diff
@@ -1460,41 +1574,49 @@ save_junit_record() {
...
-    echo "${details}" >> "${record}"
+    {
+        echo "${description}"
+        echo "${disposition}"
+        echo "${details}"
+     } >> "${record}"
 
     local tests
-    tests=$(( "$(wc -l < "${record}")" / 2 ))
+    tests=$(( "$(wc -l < "${record}")" / 3 ))
```
[the remark this answers is not in the document]
Comment (gavin-stackrox) [reply scripts/ci/lib.sh:1584]: works:
```
$ (lines=(1 2 3 4 5 6); record_length=3; lines=( "${lines[@]:${record_length}}" ); echo "${lines[*]}")
4 5 6
```

File: scripts/ci/lib.sh [hunk 14]
```diff
@@ -1339,22 +1358,109 @@ __EOM__
...
+    curl -XPOST -d @- -H 'Content-Type: application/json' "${webhook_url}" << __EOM__
+{ "text": "*An error occurred dealing with a job failure:*\n\t- Job: ${prow_job_link}.\n\t- $1." }
+__EOM__
+}
+
+_slack_check_env() {
+    (
+        set +u
+        if [[ -z "$(eval echo "\$$1")" ]]; then
+            _send_slack_error "An expected environment variable is unset/empty: $1"
```
[the remark this answers is not in the document]
Comment (gavin-stackrox) [reply scripts/ci/lib.sh:1460]: the `()` wrap takes care of that:
```
(set -u; (set +u); echo $aaa)
-bash: aaa: unbound variable
```

File: scripts/ci/lib.sh (hunk 1 above)
Comment (gavin-stackrox) [reply scripts/ci/lib.sh:913]: done

File: .openshift-ci/end.sh (hunk 2 above)
Comment (gavin-stackrox) [reply .openshift-ci/end.sh:12]: Kill joy.

Comment (gavin-stackrox) [reply .openshift-ci/end.sh:12]: done

File: .openshift-ci/end.sh (hunk 5 above)
Comment (gavin-stackrox) [reply .openshift-ci/end.sh:99]: This is _just_ a slight presentation improvement for JIRA and Slack. I don't think it is worth the effort to enumerate and maintain all the flavors (which change separately from this repo) here.

File: .openshift-ci/end.sh (hunk 6 above)
Comment (gavin-stackrox) [reply .openshift-ci/end.sh:100]: I think `# uppercase` is already sufficient. But yeah bash https://tldp.org/LDP/abs/html/xrefindex.html

File: scripts/ci/lib.sh (hunk 11 above)
Comment (gavin-stackrox) [reply scripts/ci/lib.sh:1171]: Agreed. Something to add to the improvements epic.

Comment (gavin-stackrox) [reply scripts/ci/lib.sh:1171]: https://issues.redhat.com/browse/ROX-21405

File: scripts/ci/lib.sh (hunk 14 above)
[the remark this answers is not in the document]
Comment (porridge) [reply scripts/ci/lib.sh:1460]: Ah, somehow I missed the parentheses, thanks!

File: .openshift-ci/end.sh (hunk 9 above)
Comment (porridge) [reply .openshift-ci/end.sh:141]: Hope [this link](https://www.man7.org/linux/man-pages/man1/bash.1.html#:~:text=The%20words%20between%20the%20%5B%5B%20and%20%5D%5D%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20do%20not%20undergo%20word%20splitting%20and%20pathname%20expansion.) will work in your browser. If not, search for the third hit of `[[`.

**Pull request merged**